Logging: sourcemap support for frontend stacktraces

domasx2 opened the pull request:
**What this PR does / why we need it**:
Adds source map support for logging frontend stacktraces. Will attempt on best effort basis to find sourcemaps for compiled core & plugin js bundles and indicate original source location instead of compiled file location in the stacktrace. It's much more useful to see the original source code location, as compiled files are minimized which effectively obfuscates the code. 
Using https://github.com/go-sourcemap/sourcemap. Note that it seems to fail to determine function name for react components. But not a big deal as long as file/line/column are correct. 

Before (no sourcemap usage):
![error_no_sourcemap](https://user-images.githubusercontent.com/847684/105691734-49f27100-5f06-11eb-842e-3ad112c742e5.png)

After (using sourcemap):
![error_with_sourcemap](https://user-images.githubusercontent.com/847684/105691763-51b21580-5f06-11eb-9ed5-eb65d8f33446.png)

Todos:
- [x] refactor for better testability
- [x] add unit tests
- [x] sanitize paths before attempting to read from fs?

**Which issue(s) this PR fixes**:

Related to #27993

**Special notes for your reviewer**:

Would be awesome to get opinions on a few things:

Currently all the logic is smushed together in `frontend_logging.go`. I'd like to decouple file loading, source map resolution and the actual logging and test these separately. What is the best way to go about it? Should I create an injectable `SourceMapService` and add it to `HTTPServer`, which is then passed to the frontend logging handler?

Also, current impl. keeps a cache of loaded sourcemaps, which can increase memory footprint somewhat. Is that ok? Maybe should use some cache impl that limits lifetime of the cached items, or scope the cache only to the processing of a single stacktrace?

Thanks!

domasx2 committed:
sourcemap support for error logging

domasx2 committed:
remove some unnecessary logs

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -2,17 +2,55 @@ package api
+	function string
+	line     int
+	col      int
+	pluginId string
```
```suggestion
	pluginID string
```

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -2,17 +2,55 @@ package api
+type sourceMapLocation struct {
+	dir      string
+	path     string
+	pluginId string
```
```suggestion
	pluginID string
```

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -2,17 +2,55 @@ package api
+
+type sourceMap struct {
+	consumer *sourcemap.Consumer
+	pluginId string
```
```suggestion
	pluginID string
```

aknuds1 reviewed (commented):
Just some style nits, since Go initialisms should be in consistent case.

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
 	var stacktrace = value.FmtMessage()
 	for _, frame := range value.Stacktrace.Frames {
-		stacktrace += fmt.Sprintf("\n  at %s (%s:%v:%v)", frame.Function, frame.Filename, frame.Lineno, frame.Colno)
+		mappedLocation, err := resolveSourceLocation(frame.Filename, frame.Lineno, frame.Colno)
```
Just a nit, but could you just pass on the frame? It would make for more ergonomic code, especially when the function will eventually need more or less arguments. There's no benefit to passing individual fields of `frame`.
```suggestion
		mappedLocation, err := resolveSourceLocation(frame)
```

domasx2 replied:
right! :)  
- [ ] pass the frame

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
-		stacktrace += fmt.Sprintf("\n  at %s (%s:%v:%v)", frame.Function, frame.Filename, frame.Lineno, frame.Colno)
+		mappedLocation, err := resolveSourceLocation(frame.Filename, frame.Lineno, frame.Colno)
+		if err != nil {
+			frontendLogger.Error("Error resolving stack trace frame source location.", "err", err)
```
You should return here, since you shouldn't use the other return value(s) when an error is returned.
```suggestion
			frontendLogger.Error("Error resolving stack trace frame source location.", "err", err)
			return stacktrace
```

domasx2 replied:
That would return a partial stacktrace. My intention here was to treat unexpected sorucemap error as non critical and just log the untransformed line, still better than nothing. Makes sense?

aknuds1 replied:
In that case, you should put an `else` before the next `if`. Let me see if I can make a suggestion.

domasx2 replied:
done!

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+	}
+	if strings.HasPrefix(u.Path, "/public/build/") {
+		return &sourceMapLocation{
+			dir:      setting.StaticRootPath,
```
Can you pass in a `setting.Cfg` parameter, and avoid using this global variable? We are doing our very best to get rid of global variables in the backend.

Not sure how to get `setting.Cfg` passed into your code, but that's how it should be done.

domasx2 replied:
- [ ] get `setting.Cfg` in here

domasx2 replied:
I was thinking that instead of this static location resolution here it would be better to somehow co-opt the existing static file resolution that is used for the `public/*` endpoitns. But could not figure out a way that does not require some extensive refac...

aknuds1 replied:
Dunno.

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+		if err != nil {
+			frontendLogger.Error("Error resolving stack trace frame source location.", "err", err)
+		}
+		if mappedLocation != nil {
```
Nit, but it's better to return early here, for easier reading.
```suggestion
		if mappedLocation == nil {
			return stacktrace
		}
```

domasx2 replied:
went over the code and rfaced to early returns where applicable

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+	if err != nil {
+		return nil, err
+	}
+	if smap != nil {
```
Nit, but it's better to return early to reduce nesting.
```suggestion
	if smap == nil {
		return nil, nil
	}
```

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+	}
+	if smap != nil {
+		file, function, line, col, ok := smap.consumer.Source(line, column)
+		if ok {
```
It's more idiomatic to return early.
```suggestion
		if !ok {
			return nil, nil
		}
```

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+	if strings.HasPrefix(u.Path, "/public/build/") {
+		return &sourceMapLocation{
+			dir:      setting.StaticRootPath,
+			path:     path.Join("build", u.Path[len("/public/build/"):]) + ".map",
```
Should this be a filesystem path or a network path? If it's a filesystem path, you should use `filepath.Join`.

domasx2 replied:
oops
- [ ] use `filepath.Join`

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+	if err != nil {
+		return nil, err
+	}
+	if sourceMapLocation != nil {
```
It's more idiomatic to return early here, and avoid nesting. Also wondering if you actually want to delete the cache entry?
```suggestion
	if sourceMapLocation == nil {
		delete(sourceMapCache.cache, sourceURL)
		return nil, nil
	}
```

domasx2 replied:
If there's no cache entry, and sourcemap is not found on the fs, I want to add cache entry with `nil` value so we don't try to find it in fs again later

aknuds1 replied:
OK thanks, updated my suggestion

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+	return nil, nil
+}
+
+func getSourceMap(sourceUrl string) (*sourceMap, error) {
```
In Go, initialisms should be written in consistent case.
```suggestion
func getSourceMap(sourceURL string) (*sourceMap, error) {
```

domasx2 replied:
I'll start remembering this, someday... :) 
- [ ] initialisims in all caps

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+		}
+		defer func() {
+			if err := f.Close(); err != nil {
+				frontendLogger.Error("Failed to close source map file.", "err", err)
```
We generally don't punctuate log messages:
```suggestion
				frontendLogger.Error("Failed to close source map file", "err", err)
```

domasx2 replied:
updated!

aknuds1 reviewed (commented):
Made some suggestions for more readable code mostly.

aknuds1 commented:
> Currently all the logic is smushed together in frontend_logging.go. I'd like to decouple file loading, source map resolution and the actual logging and test these separately. What is the best way to go about it? Should I create an injectable SourceMapService and add it to HTTPServer, which is then passed to the frontend logging handler?

I'm not sure. Why would you need an injectable service? Adding those is best avoided unless actually necessary, since it's a heavy abstraction and our DI framework is also fragile. I think there's only one way to figure out how to test these things in separation (if that's necessary), and it would be to try to do it. In general, try to not abstract more than necessary, since it makes the code harder to maintain.

> Also, current impl. keeps a cache of loaded sourcemaps, which can increase memory footprint somewhat. Is that ok? Maybe should use some cache impl that limits lifetime of the cached items, or scope the cache only to the processing of a single stacktrace?

I'm not sure. @marefr do you have an opinion?

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
+		if err != nil {
+			frontendLogger.Error("Error resolving stack trace frame source location.", "err", err)
+		}
+		if mappedLocation != nil {
```
I think this shouldn't be checked if `err` is non-nil, since you shouldn't trust other return values when an error is returned IMO:
```suggestion
		else if mappedLocation != nil {
```

aknuds1 reviewed (commented):
Updated a couple of suggestions.

marefr commented:
Sorry, I don't understand the full picture. Is this a problem only with plugins or in general all frontend code? Can you include some more details of log results before this PR, like errors in Grafana core itself and plugin errors.

domasx2 commented:
> Sorry, I don't understand the full picture. Is this a problem only with plugins or in general all frontend code? Can you include some more details of log results before this PR, like errors in Grafana core itself and plugin errors.

Updated description, added before/after examples of a logged stack trace, for the same error. It affects all frontend code in general, the example stack trace has code locations both from grafana core and from integrations plugin.

domasx2 commented:
> I'm not sure. Why would you need an injectable service? Adding those is best avoided unless actually necessary, since it's a heavy abstraction and our DI framework is also fragile. I think there's only one way to figure out how to test these things in separation (if that's necessary), and that's to try to do it. In general, try to not abstract more than necessary, since it makes the code harder to maintain.

My concern was that tight integration makes tests harder to write & maintain, and wasn't sure what's the accepted practice in this codebase. Testing it as a whole works for me as well, will do that then. I think I only need to abstract away the file system interaction, but hopefully will be easy to do ad-hoc

> Actually one thing that I think would make sense would be to move frontend_logging.go to a sub-package api/frontendlogging, since the api package namespace is already too populated and difficult to reason about as a result.

Makes sense!
- [ ] move to separate subpackage

aknuds1 commented:
> My concern was that tight integration makes tests harder to write & maintain, and wasn't sure what's the accepted practice in this codebase. Testing it as a whole works for me as well, will do that then. I think I only need to abstract away the file system interaction, but hopefully will be easy to do ad-hoc

Depends what you're after, but if you want to have loose coupling, you don't need DI.

On the subject of integration though, we do write integration tests and I find these easier to maintain in practice than tests that use a lot of stubbing/faking/mocking since the latter breaks down easily, and cause a lot of cognitive overhead.

domasx2 committed:
refac frontedlogging package

domasx2 committed:
make linter happy

domasx2 committed:
PR fixes

domasx2 committed:
early returns

domasx2 committed:
refac for testing & add a test

domasx2 committed:
dont punctuate logs

domasx2 commented:
> On the subject of integration though, we do write integration tests and I find these easier to maintain in practice than tests that use a lot of stubbing/faking/mocking since the latter breaks down easily, and cause a lot of cognitive overhead to figure out how (and whether) things actually fit together.

Makes sense!
I refactored everything but the http handler into `api/frontendlogging` package and made a `SourceMapStore` struct to hold config & abstraction over file access,  wrote test that tests the whole thing in one piece

domasx2 committed:
add a few comments

aknuds1 commented:
Will have a look soon, but busy with a bug at the moment.

aknuds1 commented on pkg/api/frontendlogging/source_maps.go:
```diff
@@ -0,0 +1,165 @@
+
+type ReadSourceMapFn func(dir string, path string) ([]byte, error)
+
+func ReadSourceMapFromFs(dir string, path string) ([]byte, error) {
```
Nit, but initialisms should be in consistent case in Go.
```suggestion
func ReadSourceMapFromFS(dir string, path string) ([]byte, error) {
```

domasx2 replied:
goddamit! one day...
- [ ] fix

aknuds1 commented on pkg/api/frontend_logging.go:
```diff
@@ -1,104 +1,40 @@
-func (value *frontendSentryExceptionValue) FmtMessage() string {
-	return fmt.Sprintf("%s: %s", value.Type, value.Value)
-}
+func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) frontendLogMessageHandler {
```
Is this part kept in the `api` package to avoid cyclic imports? Otherwise it would be neat to move this as well to `api/frontendlogging` so `api` can be slimmed down :)

domasx2 replied:
yup, exactly. Test utils for http handlers are in `api`. In such case they should be extracted as well, but felt like too big a scope for this PR :)

aknuds1 replied:
OK, agree then :)

aknuds1 reviewed (approved):
Looks good to me now AFAICT! Thanks for moving stuff into api/frontendlogging :)

domasx2 committed:
capitalize initialisms!

domasx2 committed:
fix path sanitization
